{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of fulltree table with ancestral and derived codons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import the data from Body and convert it into a pandas dataframe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "data = pd.read_csv(\"../../Body/3Results/fulltreeCodons.csv\", sep=\";\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's take a look at the table:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>first</th>\n",
       "      <th>second</th>\n",
       "      <th>position</th>\n",
       "      <th>ref_pos</th>\n",
       "      <th>ancestor</th>\n",
       "      <th>descendant</th>\n",
       "      <th>nuc_ref_in_ali</th>\n",
       "      <th>gene_info</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>pos_in_codon</th>\n",
       "      <th>synonymous</th>\n",
       "      <th>ancestral_aa</th>\n",
       "      <th>derived_aa</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>28409</td>\n",
       "      <td>AM_PR_0048</td>\n",
       "      <td>174</td>\n",
       "      <td>150</td>\n",
       "      <td>ttTc-</td>\n",
       "      <td>ttCc-</td>\n",
       "      <td>C</td>\n",
       "      <td>D-loop_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>non-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>28409</td>\n",
       "      <td>AM_PR_0048</td>\n",
       "      <td>593</td>\n",
       "      <td>522</td>\n",
       "      <td>acTc-</td>\n",
       "      <td>acAc-</td>\n",
       "      <td>A</td>\n",
       "      <td>D-loop_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>non-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>28409</td>\n",
       "      <td>AM_PR_0048</td>\n",
       "      <td>5007</td>\n",
       "      <td>4840</td>\n",
       "      <td>ctAac</td>\n",
       "      <td>ctGac</td>\n",
       "      <td>G</td>\n",
       "      <td>mRNA_ND2</td>\n",
       "      <td>4469.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>non-synonymous</td>\n",
       "      <td>Asn</td>\n",
       "      <td>Asp</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>28409</td>\n",
       "      <td>AM_PR_0048</td>\n",
       "      <td>7971</td>\n",
       "      <td>7774</td>\n",
       "      <td>ccGtc</td>\n",
       "      <td>ccRtc</td>\n",
       "      <td>G</td>\n",
       "      <td>mRNA_COX2</td>\n",
       "      <td>7585.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>non-synonymous</td>\n",
       "      <td>Arg</td>\n",
       "      <td>Ambiguous</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>28409</td>\n",
       "      <td>AM_PR_0048</td>\n",
       "      <td>8017</td>\n",
       "      <td>7820</td>\n",
       "      <td>ccCat</td>\n",
       "      <td>ccMat</td>\n",
       "      <td>C</td>\n",
       "      <td>mRNA_COX2</td>\n",
       "      <td>7585.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>Pro</td>\n",
       "      <td>Pro</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>577271</th>\n",
       "      <td>27068</td>\n",
       "      <td>EU_DK_0660</td>\n",
       "      <td>12579</td>\n",
       "      <td>12339</td>\n",
       "      <td>taGcc</td>\n",
       "      <td>taAcc</td>\n",
       "      <td>A</td>\n",
       "      <td>mRNA_ND5</td>\n",
       "      <td>12336.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>Stop</td>\n",
       "      <td>Stop</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>577272</th>\n",
       "      <td>27068</td>\n",
       "      <td>EU_DK_0660</td>\n",
       "      <td>12597</td>\n",
       "      <td>12357</td>\n",
       "      <td>taGcc</td>\n",
       "      <td>taAcc</td>\n",
       "      <td>A</td>\n",
       "      <td>mRNA_ND5</td>\n",
       "      <td>12336.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>Stop</td>\n",
       "      <td>Stop</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>577273</th>\n",
       "      <td>27068</td>\n",
       "      <td>EU_DK_0660</td>\n",
       "      <td>13191</td>\n",
       "      <td>12949</td>\n",
       "      <td>aaGcg</td>\n",
       "      <td>aaAcg</td>\n",
       "      <td>A</td>\n",
       "      <td>mRNA_ND5</td>\n",
       "      <td>12336.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>Lys</td>\n",
       "      <td>Lys</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>577274</th>\n",
       "      <td>27068</td>\n",
       "      <td>EU_DK_0660</td>\n",
       "      <td>15612</td>\n",
       "      <td>15366</td>\n",
       "      <td>aaCcc</td>\n",
       "      <td>aaTcc</td>\n",
       "      <td>C</td>\n",
       "      <td>mRNA_CYTB</td>\n",
       "      <td>14746.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>non-synonymous</td>\n",
       "      <td>Thr</td>\n",
       "      <td>Ile</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>577275</th>\n",
       "      <td>27068</td>\n",
       "      <td>EU_DK_0660</td>\n",
       "      <td>16588</td>\n",
       "      <td>16303</td>\n",
       "      <td>agCac</td>\n",
       "      <td>agTac</td>\n",
       "      <td>T</td>\n",
       "      <td>D-loop_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>non-coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>577276 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        first      second  position  ref_pos ancestor descendant  \\\n",
       "0       28409  AM_PR_0048       174      150    ttTc-      ttCc-   \n",
       "1       28409  AM_PR_0048       593      522    acTc-      acAc-   \n",
       "2       28409  AM_PR_0048      5007     4840    ctAac      ctGac   \n",
       "3       28409  AM_PR_0048      7971     7774    ccGtc      ccRtc   \n",
       "4       28409  AM_PR_0048      8017     7820    ccCat      ccMat   \n",
       "...       ...         ...       ...      ...      ...        ...   \n",
       "577271  27068  EU_DK_0660     12579    12339    taGcc      taAcc   \n",
       "577272  27068  EU_DK_0660     12597    12357    taGcc      taAcc   \n",
       "577273  27068  EU_DK_0660     13191    12949    aaGcg      aaAcg   \n",
       "577274  27068  EU_DK_0660     15612    15366    aaCcc      aaTcc   \n",
       "577275  27068  EU_DK_0660     16588    16303    agCac      agTac   \n",
       "\n",
       "       nuc_ref_in_ali  gene_info  gene_start  pos_in_codon      synonymous  \\\n",
       "0                   C   D-loop_2         NaN           NaN             NaN   \n",
       "1                   A   D-loop_2         NaN           NaN             NaN   \n",
       "2                   G   mRNA_ND2      4469.0           1.0  non-synonymous   \n",
       "3                   G  mRNA_COX2      7585.0           2.0  non-synonymous   \n",
       "4                   C  mRNA_COX2      7585.0           3.0      synonymous   \n",
       "...               ...        ...         ...           ...             ...   \n",
       "577271              A   mRNA_ND5     12336.0           3.0      synonymous   \n",
       "577272              A   mRNA_ND5     12336.0           3.0      synonymous   \n",
       "577273              A   mRNA_ND5     12336.0           3.0      synonymous   \n",
       "577274              C  mRNA_CYTB     14746.0           2.0  non-synonymous   \n",
       "577275              T   D-loop_1         NaN           NaN             NaN   \n",
       "\n",
       "       ancestral_aa derived_aa        note  \n",
       "0               NaN        NaN  non-coding  \n",
       "1               NaN        NaN  non-coding  \n",
       "2               Asn        Asp      normal  \n",
       "3               Arg  Ambiguous      normal  \n",
       "4               Pro        Pro      normal  \n",
       "...             ...        ...         ...  \n",
       "577271         Stop       Stop      normal  \n",
       "577272         Stop       Stop      normal  \n",
       "577273          Lys        Lys      normal  \n",
       "577274          Thr        Ile      normal  \n",
       "577275          NaN        NaN  non-coding  \n",
       "\n",
       "[577276 rows x 14 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 0. First check dataset for disrepancies (на вшивость) \n",
    "\n",
    "# ANALYSIS\n",
    "\n",
    "# 1. Simple summary descriptive statistics\n",
    "# How many different positions in codon\n",
    "\n",
    "# 2. Where along the genome\n",
    "# histogram on genome + 2000 genomes article\n",
    "# Which genes\n",
    "# are they correlated (Spearman rank and scatterplot )\n",
    "\n",
    "# 2. Loss of function or gain of function?\n",
    "# How many stop codons or not?\n",
    "# How many to stop or from stop\n",
    "\n",
    "# 3. Which aa to which\n",
    "# in which biochem group using matrices Grantham distance\n",
    "# How many ambiguous"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
